server/src/main/java/org/elasticsearch/gateway/MetaDataStateFormat.java

```diff
@@ -255,58 +255,70 @@ long findMaxStateId(final String prefix, Path... locations) throws IOException {
+                                logger.trace("found state file: {}", pav);
+                                files.add(pav);
+                            }
                         }
+                    } catch (NoSuchFileException | FileNotFoundException ex) {
+                        // no _state directory -- move on
                     }
-                } catch (NoSuchFileException | FileNotFoundException ex) {
-                    // no _state directory -- move on
                 }
             }
-        }
-        // NOTE: we might have multiple version of the latest state if there are multiple data dirs.. for this case
-        //       we iterate only over the ones with the max version.
-        long finalMaxStateId = maxStateId;
-        Collection<PathAndStateId> pathAndStateIds = files
+            // NOTE: we might have multiple version of the latest state if there are multiple data dirs.. for this case
+            //       we iterate only over the ones with the max version.
+            long finalMaxStateId = maxStateId;
+            Collection<PathAndStateId> pathAndStateIds = files
                 .stream()
                 .filter(pathAndStateId -> pathAndStateId.id == finalMaxStateId)
                 .collect(Collectors.toCollection(ArrayList::new));
 
-        final List<Throwable> exceptions = new ArrayList<>();
-        for (PathAndStateId pathAndStateId : pathAndStateIds) {
-            try {
-                T state = read(namedXContentRegistry, pathAndStateId.file);
-                logger.trace("state id [{}] read from [{}]", pathAndStateId.id, pathAndStateId.file.getFileName());
-                return state;
-            } catch (Exception e) {
-                exceptions.add(new IOException("failed to read " + pathAndStateId.toString(), e));
-                logger.debug(() -> new ParameterizedMessage(
+            final List<Throwable> exceptions = new ArrayList<>();
+            for (PathAndStateId pathAndStateId : pathAndStateIds) {
+                try {
+                    T state = read(namedXContentRegistry, pathAndStateId.file);
+                    logger.trace("state id [{}] read from [{}]", pathAndStateId.id, pathAndStateId.file.getFileName());
+                    return state;
+                } catch (NoSuchFileException | FileNotFoundException e) {
+                    // A state file was deleted by a concurrent process so the files list is not correct any longer
+                    // and we have to start over and fetch the list of state files again
+                    logger.debug(
+                        () -> new ParameterizedMessage(
+                            "{}: failed to read [{}] because it was deleted by a concurrent action, retrying...",
+                            pathAndStateId.file.toAbsolutePath(), prefix
+                        ), e
+                    );
+                    continue RETRY_LOOP;
```

Comment:
I know continuing a named loop is kinda dirty here, but with the way we collect exceptions and such it seemed like the cleanest/shortest way to add a retry.